lib/srv/db/mysql/connector.go

```diff
@@ -0,0 +1,339 @@
+	GetGCPSQLAdminClient(context.Context) (gcp.SQLAdminClient, error)
+}
+
+func (c *sslChecker) checkSSLRequired(ctx context.Context) (bool, error) {
+	if c.sometimes != nil {
+		c.sometimes.Do(func() {
+			c.requireSSL, c.requireSSLErr = checkGCPRequireSSL(ctx, c.database, c.clients)
+		})
+		return c.requireSSL, trace.Wrap(c.requireSSLErr)
+	}
+	return checkGCPRequireSSL(ctx, c.database, c.clients)
+}
+
+func checkGCPRequireSSL(ctx context.Context, database types.Database, clients gcpClients) (bool, error) {
+	gcpClient, err := clients.GetGCPSQLAdminClient(ctx)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+	// Detect whether the instance is set to require SSL.
+	// Fallback to not requiring SSL for access denied errors.
+	requireSSL, err := cloud.GetGCPRequireSSL(ctx, database, gcpClient)
+	if err != nil {
+		if trace.IsAccessDenied(err) {
+			return false, nil
+		}
+		return false, trace.Wrap(err)
+	}
+	return requireSSL, nil
+}
+
+// updateServerVersion updates the server runtime version if the version
+// reported by the database is different from the version in status
+// configuration.
+func updateServerVersion(
+	ctx context.Context,
+	log *slog.Logger,
+	db types.Database,
+	serverVersion string,
+	updateProxiedDatabase func(string, func(types.Database) error) error,
+) error {
+	statusVersion := db.GetMySQLServerVersion()
+
+	// Update only when needed.
+	if serverVersion == "" || serverVersion == statusVersion {
+		return nil
+	}
+
+	// Note that db may be a copy of the database cached by database service.
+	// Call updateProxiedDatabase to update the original as well.
+	db.SetMySQLServerVersion(serverVersion)
+	doUpdate := func(db types.Database) error {
+		db.SetMySQLServerVersion(serverVersion)
+		return nil
+	}
+
+	log.DebugContext(ctx, "Updated MySQL server version",
+		"old_version", statusVersion,
+		"version", serverVersion,
+	)
+	return trace.Wrap(updateProxiedDatabase(db.GetName(), doUpdate))
+}
+
+func newRecorderConn(conn net.Conn) *recorderConn {
+	return &recorderConn{Conn: conn}
+}
+
+type recorderConn struct {
+	net.Conn
+	buf bytes.Buffer
+}
+
+// Read implements [io.Reader]. All reads from the connection are recorded.
+func (r *recorderConn) Read(p []byte) (int, error) {
+	return io.TeeReader(r.Conn, &r.buf).Read(p)
+}
+
+// rewind sets the underlying [net.Conn] to a [bufferedConn] that reads from the
+// recorded data first, then the connection itself. After calling rewind the
+// recorder can be reused, but it should not be used once recording is no longer
+// necessary, since it buffers all reads in memory.
+// The [net.Conn] returned will replay recorded data, but it will not record
+// further reads, so it should be used once recording is no longer needed.
+func (r *recorderConn) rewind() net.Conn {
+	data := r.buf.Bytes()
+	dataCopy := make([]byte, len(data))
+	copy(dataCopy, data)
+	r.buf.Reset()
+	r.Conn = newBufferedConn(bytes.NewReader(dataCopy), r.Conn)
+	return r.Conn
+}
```

Comment:
This seems unsafe without any sort of limits. Can we end up blowing up the server memory in some scenarios?